core/types.ml

```diff
@@ -649,282 +539,305 @@ let primary_kind_of_type_arg : type_arg -> PrimaryKind.t =
+end
 
-let rec type_can_be_unl : var_set * var_set -> typ -> bool =
-  fun ((rec_vars, quant_vars) as vars) ->
-    let tcu t = type_can_be_unl vars t in
-    function
+type predicate_context = StringSet.t * var_set * var_set
+
+class virtual type_predicate = object(self)
+  method is_var : (int * subkind * freedom) -> bool = fun _ -> true
+
+  method is_point : 'a 'c . (predicate_context -> 'a -> bool) -> predicate_context -> ([< 'a meta_max_basis] as 'c) point -> bool
+    = fun f ((rec_appl, rec_vars, quant_vars) as vars) point ->
+    match Unionfind.find point with
+    | `Closed -> true
+    | `Var ((id, _, _) as var) -> IntSet.mem id quant_vars || self#is_var var
+    | `Body t -> f vars t
+    | `Recursive (var, t) -> check_rec var rec_vars true (fun rec_vars' -> f (rec_appl, rec_vars', quant_vars) t)
+
+  method is_type ((rec_appl, rec_vars, quant_vars) as vars) : typ -> bool = function
     | `Not_typed -> assert false
-    | `Effect _
-    | `Primitive _
-    | `Function _ -> true
-    | `Lolli _ -> false
-    | `Record r
-    | `Variant r -> row_can_be_unl vars r
+    | `Primitive _ -> true
+    | `Function (a, e, r) | `Lolli (a, e, r) -> self#is_type vars a && self#is_row vars e && self#is_type vars r
+    | `Record r | `Effect r | `Variant r -> self#is_row vars r
     | `Table _ -> true
     | `Lens _ -> true
-    | `Alias (_, t) -> tcu t
-    (* We might support linear lists like this...
-         but we'd need to replace hd and tl with a split operation. *)
-    (* | `Application ({Abstype.id="List"}, [`Type t]) -> tcu t *)
-    | `Application _ -> true (* TODO: change this if we add linear abstract types *)
-    | `RecursiveApplication { r_linear; _ } ->
-        (* This will have been set during desugaring, far
-         * before `type_can_be_unl` is called *)
-        not (OptionUtils.val_of (r_linear ()))
-    | `MetaTypeVar point -> point_can_be_unl type_can_be_unl vars point
-    | `ForAll (qs, t) -> type_can_be_unl (rec_vars, add_quantified_vars !qs quant_vars) t
-    | `Dual s -> type_can_be_unl vars s
-    | `End -> false
-    | #session_type -> false
-and field_can_be_unl vars =
-  function
-  | `Absent    -> true
-  | `Present t -> type_can_be_unl vars t
-  | `Var point -> point_can_be_unl field_can_be_unl vars point
-and row_can_be_unl vars (fields, row_var, _) =
-  let unl_row_var = point_can_be_unl row_can_be_unl vars row_var in
-  let unl_fields =
-    FieldEnv.fold
-      (fun _ f b -> b && field_can_be_unl vars f)
-      fields
-      true in
-  unl_row_var && unl_fields
-
-let is_unl_type = is_unl_type (IntSet.empty, IntSet.empty)
-let is_unl_row = is_unl_row (IntSet.empty, IntSet.empty)
+    | `Alias (_, t) -> self#is_type vars t
+    | `MetaTypeVar point -> self#is_point self#is_type vars point
+    | `ForAll (qs, t) -> self#is_type (rec_appl, rec_vars, add_quantified_vars !qs quant_vars) t
+    | `Application _ | `RecursiveApplication _ -> true (* For now, we assume these are acceptable! *)
```

Comment:
This seems a bit sketchy -- does it give the same behaviour as before? Why do you not need to inspect the type arguments?

Reply:
So it _technically_ gives the same behaviour as before, but that's largely because we're either rejecting (recursive) type applications, or handling them specially within the specific constraint.

That said, this behaviour is definitely not correct for `Mono`, so it's probably best to handle it :).